### DISCUSSION TITLE
Firmware auf den ESP32S3 neu aufspielen!

### DISCUSSION THREAD

**Discussion by Snusme83** [hoylabs/OpenDTU-OnBattery#1050] [General]
Servus @ all,

kann mir mal jemand erklären, wie man die Firmware OpenDTUon Battery anständig auf den Microcontroller spielt?

Hatte die Firmware schon drauf. Wollte noch ein Display (SH1106) hinzufügen. Da hab ich das pinmapping verändert und siehe da hängt sich die OpenDTUonBattery auf. Jetzt wollte ich mit Visual StudioCode die Firmware wieder aufspielen. Error - krieg sie einfach nicht drauf! Flash Erase gelingt mir auch nicht.

Kann doch nicht sein, dass ich mir jedes Mal nen neuen Mikrokontroller kaufen muss!😡

**Comment by schlimmchen** [comment]
Hast du es mal mit dem Webflasher versucht? https://solar.metacontrol.eu/opendtu-onbattery-webflasher/ Du brauchst kein VS Code nur um Firmware aufzuspielen... esptool auf der Kommandozeile reicht auch.

**Reply by Snusme83** [reply]
Da hat es mir auch immer eine Fehlermeldung angezeigt. Jetzt wollte ich ein Bildschirmfoto von der Fehlermeldung machen, jetzt funktioniert es auf einmal! Ich werde noch wahnsinnig!

Trotz alledem, ich möchte gerne alles in ein Gehäuse machen. Für das Gehäuse hätte ich ein 1,3" Display, das mit der alten openDTU ohne onBattery) bereits funktionierte.

Jetzt wenn ich in der pinmapping die GPIO von -1 auf 42 und 41 ändere, hängt sich die openDTUonBattery wahrscheinlich wieder auf und ich verbringe den ganzen Tag um es wieder zum Laufen zu kriegen!

Mit dieser pinmapping läuft es jetzt wieder:
```
[
    {
        "name": "OpenDTU ESP32-S3 + 3 x Victron MPPT",
        "nrf24": {
            "miso": 48,
            "mosi": 35,
            "clk": 36,
            "irq": 47,
            "en": 38,
            "cs": 37
        },
        "cmt": {
            "clk": -1,
            "cs": -1,
            "fcs": -1,
            "sdio": -1,
            "gpio2": -1,
            "gpio3": -1
        },
        "led": {
            "led0": -1,
            "led1": -1
        },
        "battery": {
            "rx": 43,
            "tx": 44
        },
        "victron": {
            "rx": 16,
            "tx": 15,
            "rx2": 5,
            "tx2": 4,
            "rx3": 7,
            "tx3": 6
        },
        "display": {
            "type": -1,
            "data": -1,
            "clk": -1       
        }
    }
]
```

Das ist mein ESP32S3:

![ESP32-S3GPIO](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/164622660/6e4973ed-2fb8-482a-a764-858b741a5edb)

Hab unter Display "Typ 3", "data42", "clk41" gemacht --> nach dem Einspielen der geänderten pinmapping bzw. beim auswählen des Profiles hängt sich die openDTUonBattery auf und reagiert nicht mehr!

**Reply by schlimmchen** [reply]
Ich hab ein paar Minuten draufgestarrt, aber leider hab ich keine Idee, was da schief geht...

**Reply by Snusme83** [reply]
Dass diese GPIOs (41 und 42) nicht kompatibel sind, ist nicht der Fall?

Kann es an einer Leerzeile im pinmapping liegen?

**Reply by DerHavey1994** [reply]
Ich hatte das auch mal. Du musst beim einspielen drauf achten, das du auch das pin Mapping überschreibst und nicht die config..
ggf liegt es daran

Auf meinem Bild siehst du das dort pin Mapping steht![image](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/137092915/6f676022-6e1e-477b-a73a-465a0daea563)

**Reply by schlimmchen** [reply]
Das ist ein sehr guter Hinweis. Das ist mir auch schon mind. zwei Mal passiert. Das gehört ehrlich gesagt verhindert, aber naja... Das soll dann Thomas im Upstream-Projekt machen^^

**Reply by schlimmchen** [reply]
> Dass diese GPIOs (41 und 42) nicht kompatibel sind, ist nicht der Fall?

Ich glaube, dass die für I²C verwendbar sind. Wie einfach kannst du denn andere ausprobieren? Beim ESP32-S3 gibt es keine "Input-only" GPIOs, hab ich neulich recherchiert.

> Kann es an einer Leerzeile im pinmapping liegen?

Ne, so streng ist JSON nicht.

**Reply by Snusme83** [reply]
Dass ich die config und pinmapping nicht vertauscht hab, da bin ich mir zu 99% sicher!

"Ich glaube, dass die für I²C verwendbar sind. Wie einfach kannst du denn andere ausprobieren? Beim ESP32-S3 gibt es keine "Input-only" GPIOs, hab ich neulich recherchiert."

Andere auszuprobieren wäre das geringste Problem! Allerdings scheue ich mich davor. Ich warte jetzt lieber mal bis ich die bestellten drei Ersatz-Microcontroller zuhause hab, diese gelötet sind und die Firmware drauf ist um dann ruckizucki den Microcontroller auswechseln zu können, falls das wieder so eine schwere Geburt wie heute werden soll!🤦🏻‍♂️

**Reply by Snusme83** [reply]
So, nochmal probiert, openDTUonBattery reagiert nicht mehr. Firmware neu aufspielen mit dem WebFlasher ergibt folgende Fehlermeldung:
![Screenshot Fehlermeldung Webflasher](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/164622660/4475bf1d-a746-4918-a603-15dd2a925994)

**Comment by knopers1** [comment]
kannst Du beim booten etwas auf der Konsole sehen? Ich bin gerade auch dabei die Steuerung vom ESP32 Wroom mit 4MB gegen ein ESP32-S3 16 MB umzubauen und es klappt auch ohne Probleme mit der letzten Release.... Das Image bootet und zeigt bereits 16 MB Speicher. Die NRF ist auch schon angebunden. Hast Du eventuell einige von den unten genannten GPIO´s verwendet?

Do Not Use (generally)

    gpio.43 Used for USB/Serial U0TXD
    gpio.44 Used for USB Serial U0RXD
    gpio.19 Used for native USB D-
    gpio.20 Used for native USB D+

**Reply by Snusme83** [reply]
So, jetzt läuft es wieder mit dem neuen Microcontroller (ESP32S3), nachdem ich mich ewig mit dem NRF-Modul gespielt hab, dass es verbunden ist. Den Alten habe ich jetzt wie aus geisterhand wieder mit dem Webflasher mit der Firmware bespielen können!

Zu dem Problem, dass sich der Microcontroller aufhängt/ nicht mehr reagiert:

Ich spiele folgende Pinmapping ein:

[pin_mapping.json](https://github.com/user-attachments/files/15875251/pin_mapping.json)

Das funktioniert einwandfrei. Wähle ich unter Einstellungen/ Hardware das Profil dieser pin_mapping aus, geht nix mehr!

Beim ersten Mal hatte ich für das Display type=" 3", "data=GPIO42", "clk=GPIO41" verwendet.
Beim zweiten Mal habe ich für das Display type="3", "data=GPIO2", clk=GPIO1" verwendet.

Ich hab jetzt vorerst mal die ... voll! Das ist immer ein Riesen Aufwand, das Ganze zum Laufen zu bekommen.

@schlimmchen: Die OpenDTU-OnBattery ist aber trotzdem perfekt. Werde mich erst mal von dem Gedanken verabschieden, alles in einem Gehäuse, welches ein 1,3" OLED- Display aufnehmen kann unterzubringen und das Display in Betrieb zu nehmen:-(

**Reply by schlimmchen** [reply]
Schön, dass du dich trotzdem nicht unterkriegen lässt und versuchst Spaß zu haben an diesem Projekt! Tut mir leid, dass du noch nicht dahintergestiegen bist, was dein Problem ist. Eine Sache noch: Hast du in Betracht gezogen, dass das Display defekt sein könnte? Wenn das nicht antwortet oder den I2C Bus in einem ungültigen Zustand versetzt, dann wäre es für mich absolut vorstellbar, wenn der ESP32 sich dann verklemmt. Läuft er wieder/weiter, wenn das Problem besteht und die das Display trennst?

**Reply by knopers1** [reply]
für den Pylontek in der Pinmapping hast Du GPIO 43 und 44 verwendet! Gerade das solltest Du möglicherweise nicht machen. Das könnte eventuell dein Problem sein.

``Do Not Use (generally)

gpio.43 Used for USB/Serial U0TXD
gpio.44 Used for USB Serial U0RXD
gpio.19 Used for native USB D-
gpio.20 Used for native USB D+

**Reply by schlimmchen** [reply]
@knopers1 Über GPIO19 und 20 kann ich nicht sprechen, aber 43 und 44 stehen durchaus zur freien Verfügung, falls er `generic_esp32s3_usb` verwendet, also eine Firmware, bei dem die seriellen Nachrichten über USB ausgespuckt werden.

**Reply by knopers1** [reply]
> @knopers1 Über GPIO19 und 20 kann ich nicht sprechen, aber 43 und 44 stehen durchaus zur freien Verfügung, falls er `generic_esp32s3_usb` verwendet, also eine Firmware, bei dem die seriellen Nachrichten über USB ausgespuckt werden.

das ist gut zu wissen..... Ich habe den Hinweis allgemein zum ESP32-S3 gefunden. Stand bei dem Pinout des ESP32-S3 dabei.... Deshalb habe ich gleich andere Gpio´s ausgesucht. Ich betreibe einen Victron, Pylontek, Display 1.3  und NRF+.

**Reply by Snusme83** [reply]
Ich hab das Display noch gar nicht angeschlossen!
Kann das sein, dass das das Problem ist, weil das Display noch gar nicht mit dem Microcontroller verbunden ist?🤦🏻‍♂️

**Reply by knopers1** [reply]
> Ich hab das Display noch gar nicht angeschlossen! Kann das sein, dass das das Problem ist, weil das Display noch gar nicht mit dem Microcontroller verbunden ist?🤦🏻‍♂️

nein, devinitiv nicht.

**Comment by knopers1** [comment]
@schlimmchen 
die letzte Release an dem ESP32-S3 zeigt ebenfalls eine 2MB Partition, die zu 95%  belegt ist. 

Dies wird sich sicherlich mit der nächsten Release ändern... Kannst Du sagen, wann eine neue Release für 8 MB herauskommen wird? Bis dahin betreibe ich den ESP32 Wroom mit 4 MB. Ich were bereit das neue Image zu testen.

**Reply by schlimmchen** [reply]
Nein, kann ich dir nicht sagen. Ich doktor am Power Meter schon drei Wochen länger rum, als ich dachte... Dass die Sketch-Partition zu 95% belegt ist, ist ja richtig. Die Partition ist so groß, wie sie groß ist (wie sie definiert ist) und hat mit der allgemeinen Flashspeichergröße erstmal nichts zu tun. Aber mit dem neuen Release siehst du etwas weiter oben, wie viel Flash dein ESP32 insgesamt hat.

Halte gerne die Augen offen (pull request), ich bin froh um jeden Tester, sobald das PowerMeter Refactoring getestet werden soll (das dann bereits das neue Partitionslayout nutzen wird).

**Reply by Snusme83** [reply]
Servus @ all: 
Hier nochmal die Hardwareeinstellungen und meine pinmapping im Editor:
Kann da jemand sagen, wieso es mir in der OpenDTU-OnBattery in den Hardwareeinstellungen
-Eth und
-Huawei anzeigt,
obwohl ich diese nicht im pin_mapping.json drinnen hab.

Oder ist in meiner pin_mapping.json ein Problem, weil ich Eth und Huawei nicht eingefügt hab?

Die roten Zellen sollten eigentlich auch nicht sein, oder?

Was muss ich tun, wenn ich ne andere pin_mapping.json einfüge in die OpenDTU-OnBattery und sich diese dann nach anwählen des Profiles aufhängt/ nicht mehr reagiert.

Scheinbar hab ich dann ein inkompatibles Profil gewählt. Wie soll ich dann eine Löschung über das serielle Interface durchführen?

Die OpenDTU-OnBattery ist wirklich perfekt, aber wenn ich jetzt noch mein Display anbinden könnte, wäre es nochmal perfekter:-)

![Screenshot pinmapping_hardware](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/164622660/8bf34bd4-f7a1-4f9e-b782-0ce700014dd8)
